[Windows] add system probe module for windows crash detection.

derekwbrown opened the pull request:
This PR is a replacement for previously reviewed and approved (but not merged) PR #17973 

This PR splits the previous PR into the support module implemented in system probe, and two separate forthcoming PRs.  The forthcoming PRs will each implement a new feature which uses this probe module.

Split into separate PRs to
- decouple the crash parsing module from the consumers
- allow the consumers to be added independently (and in independent order)

### Motivation

Split the previous PR into more manageable chunks.
Make it easier to review by individual code owners.
Make consumers easier to review on their own merits.

### Additional Notes


### Possible Drawbacks / Trade-offs


### Describe how to test/QA your changes

QA notes are provided in internal readme

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

derekwbrown committed:
[Windows] add system probe module for windows crash detection.

This PR is a replacement for previously reviewed and approved PR #17973.

This PR splits the previous PR into the support module implemented in
system probe, and two separate forthcoming PRs.  The forthcoming PRs
will each implement a new feature which uses this probe module.

Split into separate PRs to
- decouple the crash parsing module from the consumers
- allow the consumers to be added independently (and in independent order)

derekwbrown committed:
golint

derekwbrown committed:
golint

pgimalac commented on pkg/collector/corechecks/system/wincrashdetect/probe/README.md:
```diff
@@ -0,0 +1,114 @@
+
+### How to (really) test
+
+Unfortunately, actually testing the parser then requires actual dumps.  Actual dumps require multiple gigabytes of data.  To manually test, in a VM (don't do this on your own machine), install the crasher driver (TBD: linked here).  Start the driver. It will _immediately_ crash.  On reboot there will be a crash dump to parse, and (assuming the agent is installed) the agent will find and report.
```
`TBD: linked here`
Not sure if it's for a future PR or just forgotten

PS: you could write an e2e test with the new framework to test that !

pgimalac commented on pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:
```diff
@@ -0,0 +1,167 @@
+	for _, line := range ctx.loglines[:end] {
+
+		if strings.HasPrefix(line, debugSessionTimePrefix) {
+			parts := strings.SplitN(line, ":", 2)
+			if len(parts) == 2 {
+				wcs.DateString = strings.TrimSpace(parts[1])
+			}
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏you could use `strings.Cut` to avoid manipulating an array and checking the length:
```suggestion
			_, right, found := strings.Cut(line, ":")
			if found {
				wcs.DateString = strings.TrimSpace(right)
			}
```

pgimalac commented on pkg/process/net/check_windows.go:
```diff
@@ -0,0 +1,69 @@
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("conn request failed: socket %s, url %s, status code: %d", r.path, fmt.Sprintf(checksURL, r.path, module), resp.StatusCode)
```
```suggestion
		return nil, fmt.Errorf("conn request failed: socket %s, url %s, status code: %d", r.path, url, resp.StatusCode)
```

pgimalac reviewed (approved):
LGTM for ASC

derekwbrown committed:
Add a parsing test.

derekwbrown committed:
Rearrange code to move test-only code into test files.
(requires some add'l function call overhead).

ADd additional test.
Fix parsing problem if we received incomplete line.

derekwbrown committed:
on rebase, fix declaration

brycekahle commented:
> ## Go Package Import Differences
> Baseline: [14925fb](https://github.com/DataDog/datadog-agent/commit/14925fbb415d9de8c63ff246ac23ceef5e5883be) Comparison: [f90f680](https://github.com/DataDog/datadog-agent/commit/f90f680540ebbaa0bbe9a38a6e9ef286218460f1)
> 
> agent windows/amd64 +3, -0
> ```diff
> +github.com/DataDog/datadog-agent/pkg/collector/corechecks/system/wincrashdetect/probe
> +github.com/DataDog/datadog-agent/pkg/ebpf
> +github.com/DataDog/datadog-agent/pkg/ebpf/bytecode
> ```
> 
> agent windows/386 +3, -0
> ```diff
> +github.com/DataDog/datadog-agent/pkg/collector/corechecks/system/wincrashdetect/probe
> +github.com/DataDog/datadog-agent/pkg/ebpf
> +github.com/DataDog/datadog-agent/pkg/ebpf/bytecode
> ```
> 
> process-agent windows/amd64 +1, -0
> system-probe windows/amd64 +1, -0

I don't think anything `windows` should import `pkg/ebpf`. It looks like you call `ebpf.NewConfig()` but never actually use the config object. I think you can use the `cfg` argument passed to the module instead.

derekwbrown committed:
remove reference to ebpf.config in windows code

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:
```diff
@@ -0,0 +1,197 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+//go:build windows
+// +build windows
```
This format is no longer needed
```suggestion
```

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:
```diff
@@ -0,0 +1,197 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+//go:build windows
```
nit: add empty line to separate from license comment
```suggestion

//go:build windows
```

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:
```diff
@@ -0,0 +1,197 @@
+		line = ctx.unfinished + line
+		ctx.unfinished = ""
+	}
+	lines := strings.Split(line, "\n")
```
suggestion: `bufio.Scanner` may be able to simplify some of this newline seeking logic for you.

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:
```diff
@@ -0,0 +1,197 @@
+
+// this extra layer of indirection so that we can swap out test code which skips the actual debugger.
+func doReadCrashDump(filename string, ctx *logCallbackContext, exterr *uint32) error {
+	err := C.readCrashDump(C.CString(filename), unsafe.Pointer(ctx), (*C.long)(unsafe.Pointer(exterr)))
```
You need to free the result of `C.CString` at some point.

brycekahle replied:
Can you test your program while using `GODEBUG=cgocheck=2` as an env var? I'm unsure if the pointer passing from Go->C->Go is done correctly. You may want to use a `cgo.Handle`.

derekwbrown replied:
I ran as suggested, and the result was no error.

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go:
```diff
@@ -0,0 +1,42 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+//go:build windows
+// +build windows
```
```suggestion
```

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:
```diff
@@ -0,0 +1,124 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+//go:build windows
```
```suggestion

//go:build windows
```

brycekahle commented on pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:
```diff
@@ -0,0 +1,124 @@
+	}
+	if v > uint64(DumpTypeAutomatic) {
+		// this should never happen.  Unexpected type
+		return fmt.Errorf("Unkknown dump type %d", v)
```
```suggestion
		return fmt.Errorf("Unknown dump type %d", v)
```

brycekahle commented on pkg/process/net/check_windows.go:
```diff
@@ -0,0 +1,69 @@
+		// newly expected check names.
+	}
+	url := fmt.Sprintf(checksURL, r.path, module)
+	log.Infof("module %s trying url %s", module, url)
```
I think this log would be pretty spammy, right?

derekwbrown committed:
Review feedback

fix old use of +build vs go:build
fix memory leak using C.CString()

KevinFairise2 reviewed (approved):
LGTM.
I cannot really validate it is correct. However with the change of code owner these file should not be owned by agent-platform. I will approve it then
